src/Microsoft.AspNetCore.Sockets/Transports/LongPollingTransport.cs

```diff
@@ -34,13 +38,44 @@ public LongPollingTransport(ReadableChannel<Message> application, ILoggerFactory
                     return;
                 }
 
-                Message message;
-                if (_application.TryRead(out message))
+                // TODO: Add support for binary protocol
+                var messageFormat = MessageFormat.Text;
+                context.Response.ContentType = MessageFormatter.GetContentType(messageFormat);
+
+                var writer = context.Response.Body.AsPipelineWriter();
+                var alloc = writer.Alloc(minimumSize: 1);
+                alloc.WriteBigEndian(MessageFormatter.GetFormatIndicator(messageFormat));
+
+                while (_application.TryRead(out var message))
                 {
+                    var buffer = alloc.Memory.Span;
+
                     _logger.LogDebug("Writing {0} byte message to response", message.Payload.Length);
-                    context.Response.ContentLength = message.Payload.Length;
-                    await context.Response.Body.WriteAsync(message.Payload, 0, message.Payload.Length);
+
+                    // Try to format the message
+                    if (!MessageFormatter.TryFormatMessage(message, buffer, messageFormat, out var written))
+                    {
+                        // We need to expand the buffer
+                        // REVIEW: I'm not sure I fully understand the "right" pattern here...
+                        alloc.Ensure(MaxBufferSize);
+                        buffer = alloc.Memory.Span;
+
+                        // Try one more time
+                        if (!MessageFormatter.TryFormatMessage(message, buffer, messageFormat, out written))
+                        {
+                            // Message too large
+                            throw new InvalidOperationException($"Message is too large to write. Maximum allowed message size is: {MaxBufferSize}");
```

Comment:
Also, it this buffer **after** serializing? Because the serialized messages can be bigger than original messages in which case the user can hit a case where a message they have is actually less than the size returned in the exception message.